File: leetcode/1161/solution.py
"""
https://leetcode.com/problems/maximum-level-sum-of-a-binary-tree/
1161. Maximum Level Sum of a Binary Tree

Given the root of a binary tree, the level of its root is 1, the level of its children is 2, and so on.
Return the smallest level x such that the sum of all the values of nodes at level x is maximal.
"""
from typing import Optional, Deque
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        queue: Deque[TreeNode] = deque()
        current_level = min_level = 0
        max_value = -100001

        # Edge case
        if root is None:
            return max_value

        queue.append(root)
        while queue:
            current_level, total = current_level + 1, 0
            for _ in range(len(queue)):
                node = queue.popleft()
                total += node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            if max_value < total:
                max_value, min_level = total, current_level

        return min_level

File: leetcode/1161/solution.rs
// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

impl Solution {
    pub fn max_level_sum(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut smallest_level = 1;
        let mut max_total = i32::MIN;
        let mut queue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::new();
        queue.push_back(root.unwrap());

        let mut num_nodes = queue.len();
        let mut current_level = 1;
        let mut current_total = 0;
        while let Some(node) = queue.pop_front() {
            num_nodes -= 1;
            let node = node.borrow();
            current_total += node.val;
            if let Some(node) = &node.left {
                queue.push_back(Rc::clone(node));
            }
            if let Some(node) = &node.right {
                queue.push_back(Rc::clone(node));
            }
            if num_nodes == 0 {
                // End of level
                if max_total < current_total {
                    max_total = current_total;
                    smallest_level = current_level;
                }
                num_nodes = queue.len();
                current_level += 1;
                current_total = 0;
            }
        }

        smallest_level
    }
}
